File: backend/app/schemas/board.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, EmailStr

from app.models.enums import BoardRole, BoardVisibility


class BoardBase(BaseModel):
    name: str
    description: str | None = None
    background: str | None = None
    visibility: BoardVisibility = BoardVisibility.workspace


class BoardCreate(BoardBase):
    workspace_id: uuid.UUID


class BoardUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    background: str | None = None
    visibility: BoardVisibility | None = None
    is_archived: bool | None = None


class BoardRead(BoardBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    workspace_id: uuid.UUID
    is_archived: bool
    is_template: bool
    created_by: uuid.UUID
    created_at: datetime


class BoardMemberRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    board_id: uuid.UUID
    user_id: uuid.UUID
    role: BoardRole
    is_starred: bool


class BoardMemberCreate(BaseModel):
    email: EmailStr
    role: BoardRole = BoardRole.member


class BoardMemberUpdate(BaseModel):
    role: BoardRole


class BoardStarUpdate(BaseModel):
    is_starred: bool

File: frontend/src/types/board.ts
export type BoardRole = "admin" | "member" | "viewer";
export type BoardVisibility = "private" | "workspace" | "public";

export interface Board {
  id: string;
  name: string;
  description: string | null;
  background: string | null;
  visibility: BoardVisibility;
  workspace_id: string;
  is_archived: boolean;
  is_template: boolean;
  created_by: string;
  created_at: string;
}

export interface BoardCreate {
  name: string;
  description?: string | null;
  background?: string | null;
  visibility?: BoardVisibility;
  workspace_id: string;
}

export interface BoardUpdate {
  name?: string;
  description?: string | null;
  background?: string | null;
  visibility?: BoardVisibility;
  is_archived?: boolean;
}

export interface BoardMember {
  id: string;
  board_id: string;
  user_id: string;
  role: BoardRole;
  is_starred: boolean;
}

export interface BoardMemberCreate {
  email: string;
  role?: BoardRole;
}

export interface BoardMemberUpdate {
  role: BoardRole;
}

export interface BoardInviteLink {
  id: string;
  board_id: string;
  token: string;
  created_at: string;
}

export interface BoardJoinResult {
  board_id: string;
  already_member: boolean;
}
